### PR TITLE
Add convenience functions to set preferences

### PR DESCRIPTION
**Pull request by JoshuaLampert**
Adds the functions `set_library_p4est!` and `set_library_sc!` to conveniently set the preferences for custom libraries.

### CODE REVIEW

**Comment by JoshuaLampert** [comment]
Looks like the new P4est_jll destroyed something :grimacing:
```julia
julia> using Pkg

julia> Pkg.add(["MPIPreferences", "P4est_jll", "MPI"])
[...]
julia> using MPIPreferences

julia> MPIPreferences.use_system_binary()
┌ Info: MPI implementation identified
│   libmpi = "libmpi"
│   version_string = "Open MPI v4.1.0, package: Debian OpenMPI, ident: 4.1.0, repo rev: v4.1.0, Dec 18, 2020\0"
│   impl = "OpenMPI"
│   version = v"4.1.0"
└   abi = "OpenMPI"
┌ Info: MPIPreferences changed
│   binary = "system"
│   libmpi = "libmpi"
│   abi = "OpenMPI"
│   mpiexec = "mpiexec"
│   preloads = Any[]
└   preloads_env_switch = nothing
```
and after restarting:
```julia
julia> using P4est_jll
julia> using MPI
[ Info: Precompiling MPI [da04e1cc-30fd-572f-bb4f-1f8673147195]

julia> MPI.Init()
--------------------------------------------------------------------------
It looks like opal_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during opal_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  opal_shmem_base_select failed
  --> Returned value -1 instead of OPAL_SUCCESS
--------------------------------------------------------------------------
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  opal_init failed
  --> Returned value Error (-1) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
--------------------------------------------------------------------------
It looks like MPI_INIT failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during MPI_INIT; some of which are due to configuration or environment
problems.  This failure appears to be an internal failure; here's some
additional information (which may only be relevant to an Open MPI
developer):

  ompi_mpi_init: ompi_rte_init failed
  --> Returned "Error" (-1) instead of "Success" (0)
--------------------------------------------------------------------------
*** An error occurred in MPI_Init_thread
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
[saola6:24126] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
```
Without `using P4est_jll` it works (of course) and the same, but with P4est_jll v2.8.1.+2, too. I think that's the problem behind at least some of the CI failures. Any ideas?

**Comment by JoshuaLampert** [comment]
It's probably because of a mismatch between the OpenMPI versions (my local one and the one that is used in the OpenMPI build of P4est_jll), right?

**Comment by JoshuaLampert** [comment]
Ok, the same happens with HDF5_jll. So that's probably not the problem...

File: src/P4est.jl [hunk 1]
```diff
@@ -45,6 +46,45 @@ to check whether `p4est` has been initialized.
...
 package_id() = unsafe_load(cglobal((:p4est_package_id, LibP4est.libp4est), Cint))
 
+const P4EST_UUID = UUID("7d669430-f675-4ae7-b43e-fab78ec5a902")
+
+"""
+    P4est.set_library_p4est!(path; force = true)
+
+Set the `path` to a system-provided `p4est` installation. Restart the Julia session
+after executing this function so that the changes take effect. Calling this
+function is necesarry, when you want to use a system-provided `p4est`
```
Comment (ranocha) [inline src/P4est.jl:56]:
```suggestion
function is necessary when you want to use a system-provided `p4est`
```

File: src/P4est.jl [hunk 2]
```diff
@@ -45,6 +46,45 @@ to check whether `p4est` has been initialized.
...
+    end
+    @info "Please restart Julia and reload P4est.jl for the library changes to take effect"
+end
+
+"""
+    P4est.set_library_sc!(path; force = true)
+
+Set the `path` to a system-provided `sc` installation. Restart the Julia session
+after executing this function so that the changes take effect. Calling this
+function is necesarry, when you want to use a system-provided `p4est`
```
Comment (ranocha) [inline src/P4est.jl:74]:
```suggestion
function is necessary, when you want to use a system-provided `p4est`
```

File: src/P4est.jl [hunk 3]
```diff
@@ -45,6 +46,45 @@ to check whether `p4est` has been initialized.
...
+installation on Windows or when you want to use a another `sc`
```
Comment (ranocha) [inline src/P4est.jl:75]:
```suggestion
installation on Windows or when you want to use another `sc`
```

**Comment by ranocha** [comment]
There are some really weird issues:

```
mpiexec: Error: unknown option "-n"
```
for [ubuntu-latest - P4EST_CUSTOM_MPI_CUSTOM - Julia 1](https://github.com/trixi-framework/P4est.jl/pull/88#logs) and

```
  ✗ P4est
  19 dependencies successfully precompiled in 13 seconds

ERROR: The following 1 direct dependency failed to precompile:

P4est [7d669430-f675-4ae7-b43e-fab78ec5a902]

Failed to precompile P4est [7d669430-f675-4ae7-b43e-fab78ec5a902] to "C:\\Users\\runneradmin\\.julia\\compiled\\v1.9\\P4est\\jl_16C0.tmp".

[1108] signal (22): SIGABRT
in expression starting at D:\a\P4est.jl\P4est.jl\src\LibP4est.jl:10
```
for [windows-latest - P4EST_JLL_MPI_DEFAULT - Julia 1](https://github.com/trixi-framework/P4est.jl/pull/88#logs)

**Comment by JoshuaLampert** [comment]
Looks like P4est_jll doesn't use the right MPI version. After setting all Preferences (MPIPreferences and P4est preference) and restarting the REPL on Roci:
```julia
julia> using P4est

shell> ompi_info
                 Package: Debian OpenMPI
                Open MPI: 4.0.3
  Open MPI repo revision: v4.0.3
   Open MPI release date: Mar 03, 2020
                Open RTE: 4.0.3
  Open RTE repo revision: v4.0.3
   Open RTE release date: Mar 03, 2020
                    OPAL: 4.0.3
      OPAL repo revision: v4.0.3
       OPAL release date: Mar 03, 2020
                 MPI API: 3.1.0
            Ident string: 4.0.3
                  Prefix: /home/jlampert/.julia/artifacts/6f1138d9f7bcc8575ac98bb3ccbc47505c718c80
 Configured architecture: x86_64-pc-linux-gnu
          Configure host: lcy01-amd64-020
           Configured by: buildd
           Configured on: Wed Apr 15 13:14:35 UTC 2020
          Configure host: lcy01-amd64-020
[...]
```
which means that it still uses the OpenMPI version from the Artifact and not the system OpenMPI version. On the other hand after performing the same steps with HDF5:
```julia
julia> using HDF5

shell> ompi_info
                 Package: Debian OpenMPI
                Open MPI: 4.0.3
  Open MPI repo revision: v4.0.3
   Open MPI release date: Mar 03, 2020
                Open RTE: 4.0.3
  Open RTE repo revision: v4.0.3
   Open RTE release date: Mar 03, 2020
                    OPAL: 4.0.3
      OPAL repo revision: v4.0.3
       OPAL release date: Mar 03, 2020
                 MPI API: 3.1.0
            Ident string: 4.0.3
                  Prefix: /usr
 Configured architecture: x86_64-pc-linux-gnu
          Configure host: lcy01-amd64-020
           Configured by: buildd
           Configured on: Wed Apr 15 13:14:35 UTC 2020
          Configure host: lcy01-amd64-020
[...]
```
With HDF5.jl `ompi_info` reports to use the system MPI version located under `/usr` after setting the preference for the custom hdf5 library. In contrast, setting the preference for the p4est library doesn't change the MPI version that is used. I guess that is the problem at least for the failure on Ubuntu. https://github.com/open-mpi/ompi/issues/4557 indicates that the error message `mpiexec: Error: unknown option "-n"` is caused by a conflict of OMPI installations.

**Comment by ranocha** [comment]
The kind of failure

```
  ✗ P4est
  19 dependencies successfully precompiled in 13 seconds

ERROR: The following 1 direct dependency failed to precompile:

P4est [7d669430-f675-4ae7-b43e-fab78ec5a902]

Failed to precompile P4est [7d669430-f675-4ae7-b43e-fab78ec5a902] to "C:\\Users\\runneradmin\\.julia\\compiled\\v1.9\\P4est\\jl_16C0.tmp".

[1108] signal (22): SIGABRT
in expression starting at D:\a\P4est.jl\P4est.jl\src\LibP4est.jl:10
```
for [windows-latest - P4EST_JLL_MPI_DEFAULT - Julia 1](https://github.com/trixi-framework/P4est.jl/pull/88#logs)


also happens in our CI tests of Trixi.jl. Do you have an idea what's causing this? If not, I would suggest rolling back the PR to P4est_jll to hotfix CI for now

**Comment by sloede** [comment]
We're seeing other, likely p4est-related issues also coming up for libtrixi (e.g., [here](https://github.com/trixi-framework/libtrixi/actions/runs/6452511161/job/17514658132?pr=134#step:14:516)). I wonder if something is severely broken with the latest JLL, though I can't understand what and why.

However, if this persists, we might need to consider to reverting the JLL and stripping OpenMPI from its build options, then fix it first locally, before recreating the JLL with OpenMPI enabled.

**Comment by ranocha** [comment]
Yes, please. Can we proceed as follows to fix this?

1. Revert the P4est_jll build to make everything work again for now
2. Add explicit compat bopunds for P4est_jll.jl to P4est.jl, allowing only versions that we have tested (CompatHelper can help us updating if necessary)
3. Try to debug the P4est_jll build
4. Update P4est.jl to a new P4est_jll if possible

**Review by ranocha — APPROVED** [review]
Thanks!

**Pull request merged**